### PR TITLE
Add code for Inventory Module with more update operations

### PR DESCRIPTION
**Pull request by Abhishek-121**
### Added  code for 
- Update Device role, 
- update interface details, 
- create and assign Global User defined field to devices, 
- Delete Global UDF, 
- Delete Provisioned/Unprovisioned devices, 
- convert playbook variable into snakecase, 
- add examples and docstring to each API.
- Added playbook for inventory operation

### CODE REVIEW

File: plugins/modules/inventory_intent.py [hunk 1]
```diff
@@ -661,20 +899,56 @@ def get_diff_merged(self, config):
...
+                self.status = "failed"
+                self.result['changed'] = False
+                return self
+
+            # Now add code for adding Global UDF to device with Id
+            self.add_field_to_devices(device_ids).check_return_status()
+
+            self.result['changed'] = True
+            log_msg = "Global User Defined Added with name {0} added to device Successfully !".format(field_name)
+            log(log_msg)
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:935]: can we change to self.log?

File: plugins/modules/inventory_intent.py [hunk 2]
```diff
@@ -713,59 +985,232 @@ def get_diff_merged(self, config):
...
+                if device in self.have.get("device_in_dnac"):
+                    device_present = True
+                    break
+
+            if not device_present:
+                msg = "Cannot perform Update operation as device - {0} not present in Cisco DNA Center".format(str(device_to_update))
+                self.status = "success"
+                self.result['changed'] = False
+                self.result['response'] = msg
+                log(msg)
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1006]: can we change to self.log

File: plugins/modules/inventory_intent.py [hunk 3]
```diff
@@ -713,59 +985,232 @@ def get_diff_merged(self, config):
...
+                            'interface_uuid': interface_id,
+                            'deployment_mode': interface_params.get('deployment_mode', 'Deploy')
+                        }
+                        response = self.dnac._exec(
+                            family="devices",
+                            function='update_interface_details',
+                            op_modifies=True,
+                            params=update_interface_params,
+                        )
+                        log(str(response))
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1101]: can we change to self.log?

File: plugins/modules/inventory_intent.py [hunk 4]
```diff
@@ -713,59 +985,232 @@ def get_diff_merged(self, config):
...
+                    except Exception as e:
+                        error_message = "Error while Updating Interface Details in Cisco DNA Center - {0}".format(str(e))
+                        self.status = "success"
+                        self.result['changed'] = False
+                        log_msg = "Port actions are only supported on user facing/access ports as it's not allowed or No Updation required"
+                        self.result['log'] = log_msg
+
+                return self
+
+            else:
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1130]: as we are returning from if, do we need this else?

File: plugins/modules/inventory_intent.py [hunk 5]
```diff
@@ -713,59 +985,232 @@ def get_diff_merged(self, config):
...
+                msg = "Devices {0} present in Cisco DNA Center and updated successfully".format(config['ip_address'])
+                self.log(msg)
+                self.result['log'].append(msg)
+                self.status = "success"
 
-            log(str(response))
-            device_added = True
+            return self
+
+        else:
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1174]: this else also not required. we can restructure the code

File: plugins/modules/inventory_intent.py [hunk 6]
```diff
@@ -713,59 +985,232 @@ def get_diff_merged(self, config):
...
+            config['ip_address'] = devices_to_add
+            self.mandatory_parameter().check_return_status()
+            try:
+                response = self.dnac._exec(
+                    family="devices",
+                    function='add_device',
+                    op_modifies=True,
+                    params=self.want.get("device_params"),
+                )
+                log(str(response))
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:1185]: can you please change to self.log?

File: playbooks/inventory_device.yml [hunk 7]
```diff
@@ -19,21 +19,21 @@
...
+          ip_address: "{{item.ip_address}}"
+          snmp_auth_passphrase: "{{item.snmp_auth_passphrase}}"
+          snmp_auth_protocol: "{{item.snmp_auth_protocol}}"
+          snmp_mode: "{{item.snmp_mode}}"
+          snmp_priv_passphrase: "{{item.snmp_priv_passphrase}}"
+          snmp_priv_protocol: "{{item.snmp_priv_protocol}}"
+          snmp_ro_community: "{{item.snmp_ro_community}}"
+          snmp_rw_community: "{{item.snmp_rw_community}}"
+          snmp_username: "{{item.snmp_username}}"
+          username: "{{item.username}}"
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply playbooks/inventory_device.yml:33]: moved

File: plugins/modules/inventory_intent.py [hunk 8]
```diff
@@ -35,109 +35,148 @@
...
             type: str
-          newMgmtIpAddress:
-            description: Network Device's newMgmtIpAddress.
+          new_mgmt_ipaddress:
+            description: Device's new Mgmt IpAddress.
             type: str
-        type: list
-      userName:
-        description: Network Device's userName.Required for Adding Network Device.
+      username:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:138]: moved

File: plugins/modules/inventory_intent.py [hunk 9]
```diff
@@ -470,11 +588,95 @@ def device_exists_in_dnac(self, want_device):
...
+            self.log(str(response))
+            response = response.get("response")
+            field_name = self.config[0].get('add_user_defined_field').get('name')
+            log("Global User Defined Field with name - {0} created successfully").format(field_name)
+
+        except Exception as e:
+            error_message = "Error while Creating Global User Defined Field in Cisco DNA Center - {0}".format(str(e))
+            log(error_message)
+
+        return self
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:627]: set status as success.

File: plugins/modules/inventory_intent.py [hunk 10]
```diff
@@ -470,11 +588,95 @@ def device_exists_in_dnac(self, want_device):
...
+        if (len(udf) == 1):
+            udf_exists = True
+        else:
+            message = "Global User Defined Field with name - {0} doesnot exist in Cisco DNA Center".format(field_name)
+            self.log(message)
+
+        return udf_exists
+
+    def create_user_defined_field(self):
+        try:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:611]: Docstring added.

File: plugins/modules/inventory_intent.py [hunk 11]
```diff
@@ -470,11 +588,95 @@ def device_exists_in_dnac(self, want_device):
...
+            payload = self.config[0].get('add_user_defined_field')
+            response = self.dnac._exec(
+                family="devices",
+                function='create_user_defined_field',
+                params=payload,
+            )
+            self.log(str(response))
+            response = response.get("response")
+            field_name = self.config[0].get('add_user_defined_field').get('name')
+            log("Global User Defined Field with name - {0} created successfully").format(field_name)
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:621]: changed.

File: plugins/modules/inventory_intent.py [hunk 12]
```diff
@@ -470,11 +588,95 @@ def device_exists_in_dnac(self, want_device):
...
+                log(str(response))
+                response = response.get("response")
+
+            except Exception as e:
+                error_message = "Error while Adding Global UDF to device in Cisco DNA Center - {0}".format(str(e))
+                log(error_message)
+                self.status = "failed"
+                self.result['changed'] = False
+
+        return self
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:656]: add status as success if we reecive the api response.

File: plugins/modules/inventory_intent.py [hunk 13]
```diff
@@ -470,11 +588,95 @@ def device_exists_in_dnac(self, want_device):
...
+                function='get_all_user_defined_fields',
+                params={"name": field_name},
+            )
+            log(str(response))
+            udf = response.get("response")
+            udf_id = udf[0].get("id")
+
+            return udf_id
+
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:671]: moved and if we don't receive any value from this API then we are checking things from where this api being called.

File: plugins/modules/inventory_intent.py [hunk 14]
```diff
@@ -483,26 +685,28 @@ def mandatory_parameter(self, config):
...
         if mandatory_params_absent:
             self.msg = "Mandatory paramters {0} not present".format(mandatory_params_absent)
             self.result['msg'] = "Required parameters {0} for adding devices are not present".format(mandatory_params_absent)
             self.status = "failed"
         else:
-            self.result['msg'] = "Required paramter for Adding the devices in Inventory are present."
-            self.msg = "Required paramter for Adding the devices in Inventory are present."
+            log_msg = "Required paramter for Adding the devices in Inventory are present."
+            self.result['msg'] = log_msg
+            self.msg = log_msg
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:709]: we can, changed as per comment.

File: plugins/modules/inventory_intent.py [hunk 15]
```diff
@@ -612,17 +827,40 @@ def get_device_ids(self, device_ips):
...
             except Exception as e:
                 error_message = "Error while fetching device from Cisco DNA Center - {0}".format(str(e))
-                self.log(error_message)
+                log(error_message)
                 raise Exception(error_message)
 
         return device_ids
 
+    def get_interface_from_ip(self, device_ip):
+
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:844]: added.

File: plugins/modules/inventory_intent.py (hunk 1 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:935]: changed.

File: plugins/modules/inventory_intent.py (hunk 3 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1101]: changed.

File: plugins/modules/inventory_intent.py (hunk 5 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1174]: Make the required changes.

File: plugins/modules/inventory_intent.py (hunk 4 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:1130]: chnaged.

File: plugins/modules/inventory_intent.py [hunk 16]
```diff
@@ -470,11 +588,95 @@ def device_exists_in_dnac(self, want_device):
 
         return device_in_dnac
 
-    def mandatory_parameter(self, config):
+    def is_udf_exist(self, field_name):
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:591]: added docstring and modify the function.

**Review by madhansansel — APPROVED** [review]

**Pull request merged**